### ISSUE TITLE
Airflow 3.0 deprecations

### ISSUE THREAD

**Issue by dhruvmanila** [astral-sh/ruff#14626]
Tracking issue for the `AIR3xx` rules that implements Airflow 3.0 deprecations.

Ref: https://github.com/apache/airflow/issues/41641

**Comment by Lee-W** [comment]
There's already a short proposal https://github.com/astral-sh/ruff/pull/14616#issuecomment-2503207647. 

---

I'm thinking of separating "removal" and "rename" at least. It would be much easier for the users to know whether there's anything they could do to fix it.

**Comment by dhruvmanila** [comment]
Additionally, the [NumPy 2.0 migration rule](https://docs.astral.sh/ruff/rules/numpy2-deprecation/) is a good reference whose implementation is at: https://github.com/astral-sh/ruff/blob/6f1cf5b686e67fb74383641fcbbe98e33481a271/crates/ruff_linter/src/rules/numpy/rules/numpy_2_0_deprecation.rs

**Comment by Lee-W** [comment]
Hi, we have grouped our deprecations into several categories and would appreciate it if the team could take a look and let us know what you think about it 🙂


* already merged
    * AIR301: Avoid implicit DAG schedule https://github.com/astral-sh/ruff/pull/14581
* newly proposed error number
    * AIR302: removal or no direct mapping
        * following https://github.com/astral-sh/ruff/pull/14582 
    * AIR303: rename
        * It would be much easier for the users to know whether they could do anything to fix it.
    * AIR304: moved to airflow providers
        *  for these, they're similar to "rename" but require the user to install an additional package (e.g., `apache-airflow-providers-fab`)

Here's the categorised list we have now
https://github.com/apache/airflow/issues/41641#issuecomment-2508140120
(there's another group we temporarily mark as AIR305, but we'll need some time to figure out what we can do for those. Will likely to be merged into 302 or 303)

**Comment by MichaReiser** [comment]
I like what's proposed here. Thanks for going through all the deprecations and grouping them by possible rules. We can also do some fine-tuning as we go.

**Comment by Lee-W** [comment]
Later discuss with @uranusjr, we think it would be better for us to merge rename and removal into `302`. So the current idea would be 

* already merged
    * AIR301: Avoid implicit DAG schedule (https://github.com/astral-sh/ruff/pull/14581)
    * AIR302: removal (https://github.com/astral-sh/ruff/pull/14582)
* new rule number
    * AIR303: moved to airflow providers
        * for these, they're similar to "rename" but require the user to install an additional package (e.g., apache-airflow-providers-fab)

We now use https://github.com/apache/airflow/issues/44556 to track the detail of this task instead of the original one as some of the migration might not be able to be done in ruff

**Comment by MichaReiser** [comment]
Sorry for pinging you here, but could someone more familiar with Airflow volunteer looking at the changes in https://github.com/astral-sh/ruff/pull/14631? We just want to make sure that the change is correct.

**Comment by Lee-W** [comment]
> Sorry for pinging you here, but could someone more familiar with Airflow volunteer looking at the changes in [#14631](https://github.com/astral-sh/ruff/pull/14631)? We just want to make sure that the change is correct.

Looks like a valid one to me

**Comment by MichaReiser** [comment]
Thanks for taking a look!

**Comment by dhruvmanila** [comment]
I think it'll also be really useful to provide an auto-fix for these rules. Refer to how NumPy 2.0 deprecation rule does it:

https://github.com/astral-sh/ruff/blob/1bd8fbb6e862259a4515f180c2b2ab119d3ffc23/crates/ruff_linter/src/rules/numpy/rules/numpy_2_0_deprecation.rs#L683-L708

And, I think we'll also need to add this guard check specific to Airflow imports and attributes which allows for having both the new and deprecated symbols to exists in a `try .. except ..` block:

https://github.com/astral-sh/ruff/blob/1bd8fbb6e862259a4515f180c2b2ab119d3ffc23/crates/ruff_linter/src/rules/numpy/rules/numpy_2_0_deprecation.rs#L671-L673

**Comment by uranusjr** [comment]
I agree. We’re on a tight schedule and would like to prioritise adding all the rules first, but contribution is always welcomed on these.

**Comment by dhruvmanila** [comment]
> We’re on a tight schedule and would like to prioritise adding all the rules first, but contribution is always welcomed on these.

Yeah, I think that's fine. Do you have a rough timeline on when v3 is going to be released? I could spend some time in adding the infrastructure for an auto-fix. I think that shouldn't take much time as we already have a reference implementation. And, I'll mark this as a "help wanted" task in the issue description if someone else is interested.

**Comment by tirkarthi** [comment]
It seems ruff also has interface to add fixes which seems straightforward. I tested below where users can opt for unsafe fixes using `--fix` and also see `--diff` before applying them. This fixes some of the common keyword argument related issues. I can raise a PR. This will help users to do migrations using ruff with in place changes for the dag files once they confirm the changes through diffs in large deployments. Sample output as below 

Diff : https://github.com/astral-sh/ruff/compare/main...tirkarthi:ruff:air302-fixes

```
cargo run -p ruff -- check --preview --select=AIR302 --unsafe-fixes --diff example_latest_only.py --no-cache
   Compiling ruff v0.8.2 (/home/karthikeyan/stuff/python/ruff/crates/ruff)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 11.86s
     Running `target/debug/ruff check --preview --select=AIR302 --unsafe-fixes --diff example_latest_only.py --no-cache`
--- example_latest_only.py
+++ example_latest_only.py
@@ -27,7 +27,7 @@
 
 with DAG(
     dag_id="latest_only",
-    schedule_interval="daily",
+    schedule="daily",
     start_date=datetime.datetime(2021, 1, 1),
     catchup=False,
     tags=["example2", "example3"],

Would fix 1 error.
```

**Comment by dhruvmanila** [comment]
@tirkarthi Thank you for working on that. I think it will be useful to consider the kind of deprecations and provide the auto-fixes accordingly. Your implementation looks simple enough, I'm happy to look at the PR. One suggestion for the `None` case would be to use the [`remove_argument`](https://github.com/astral-sh/ruff/blob/e3f34b8f5bd1eb715e732e5a277aacedac17358b/crates/ruff_linter/src/fix/edits.rs#L201-L212) function as there's no replacement for that. We could even rename `Replacement::None` to `Replacement::Remove` to make the intentions clear.

**Comment by tirkarthi** [comment]
Thanks @dhruvmanila, created PR #14887 . I agree that the fixes need to be categorized like numpy with `Replacement` enum updated with more details but I guess it will be a bigger effort and will see if I can tackled it in a separate PR.

**Comment by Lee-W** [comment]
Just created a new one for AIR303 https://github.com/astral-sh/ruff/pull/14764

**Comment by tirkarthi** [comment]
Is there a helper method to change the import path while keeping the symbol name? I looked at numpy replacements and they seem to do something like below where the import path remains the same and new symbol is introduced. In terms of Airflow deprecations the symbol remains the same but import path changes. Just changing the import path of the replacement through `checker.importer().get_or_import_symbol` causes below type of code to be generated I guess and `ConflictingName` is raised.

```python
from airflow.operators.empty import EmptyOperator
from airflow.providers.standard.operators.empty import EmptyOperator
```

Deprecated

```python
from numpy import all
```

Fix

```python
from numpy import all, alltrue
```

Deprecated and removed that will cause `ImportError`

```python
from airflow.operators.empty import EmptyOperator
```

Fix

```python
from airflow.providers.standard.operators.empty import EmptyOperator
```

**Comment by MichaReiser** [comment]
@Lee-W and @uranusjr What's the status of the airflow rules? Or, we're working on the Ruff 0.10 minor and we noticed that AIR301 would meet our stabilization criteria but I don't think it makes sense to only stabilize some of them. What do you think?

**Comment by Lee-W** [comment]
I don't think we'll add new things to AIR301, but there're still some for AIR302 and AIR303 that we're currently out of bandwidth to work on

**Comment by MichaReiser** [comment]
Sounds reasonable. Thanks for the update.

**Comment by Lee-W** [comment]
Hey @dhruvmanila @MichaReiser,

Recently, we've been testing these `AIR` rules with some existing dags to see how they work. However, we discovered there are too many warnings for even simple cases. Some might not be necessary now (backward compat layers have been added.) In addition to adding an auto fix, we'd also like to propose a reorg of these error codes. It would make airflow users migrate from Airflow 2 to Airflow 3 without so much pressure.

## Original

* AIR001
    * Name: airflow-variable-name-task-id-mismatch
    * Message: Task variable name should match the task_id: "{task_id}"
    * Status: 
* AIR301
    * Name: airflow-dag-no-schedule-argument
    * Message: DAG should have an explicit schedule argument
    * Status: 🧪
* AIR302
    * Name: airflow3-removal
    * Message: {deprecated} is removed in Airflow 3.0
    * Status: 🧪 🛠️
* AIR303
    * Name: airflow3-moved-to-provider
    * Message: {deprecated} is moved into {provider} provider in Airflow 3.0;
    * Status: 🧪

## Proposal

* AIR001 (**not changed**)
    * Name: airflow-variable-name-task-id-mismatch
    * Message: Task variable name should match the task_id: "{task_id}"
    * Status: 
* AIR201 (**original AIR301**)
    * Name: airflow-dag-no-schedule-argument
    * Message: DAG should have an explicit schedule argument
    * Note: https://github.com/astral-sh/ruff/pull/16978
* AIR302
    * Name: airflow3-removal
    * Import or usage removed without fallback in Airflow 3.0 
    * Note:
        * e.g., `airflow.operators.subdag`
        * Some existing rules should be moved to `AIR304`
* AIR303
    * Name: airflow3-moved-to-provider
    * Message: Import or usage moved from airflow to `airflow.providers` in Airflow 3.0 with no fallback
    * Note: 
        * e.g., `airflow.hooks.zendesk_hook.ZendeskHook` → `airflow.providers.zendesk.hooks.zendesk.ZendeskHook`
        * Some existing rules should be moved to `AIR305`
* AIR304
    * Name: airflow3-0-deprecated
    * Message: The feature is deprecated in 3.0 and will be removed in a future version (e.g., 3.1 or 4.0).
    * Note:
        * e.g., `airflow.datasets.Dataset` → `airflow.sdk.Asset` (it's expected to be moved in 3.2)
* AIR305
    * Name: airflow3-0-deprecated-and-moved-to-provider 
    * Message: Feature is deprecated in 3.0 and will move from airflow to airflow.providers in a future version
    * Notes:
        * e.g., `airflow.triggers.temporal` → `airflow.providers.standard.triggers.temporal`

---

For those who want to migrate to Airflow 3.0, `--select AIR302,AIR303` will be required, and we intend to make this set much smaller. For aggressive people, they can turn on `--select AIR3`. `AIR305` is one thing I'm not that sure but I will discuss it with the community and get back to you. Also, would like to know if we move `AIR301`, do we need to reindex other `AIR3xx` as well?

**Comment by Lee-W** [comment]
Hey @MichaReiser @dhruvmanila, just a gentle ping in caseyou missed the message. 🙂

I also created some auto fixing PR https://github.com/astral-sh/ruff/pulls/Lee-W . Would be awesome if I could get some suggestions. Thanks! 🙏

**Comment by dhruvmanila** [comment]
> Hey [@MichaReiser](https://github.com/MichaReiser) [@dhruvmanila](https://github.com/dhruvmanila), just a gentle ping in case you missed the message. 🙂

Thank you for the ping! We've been busy this week with the company on-site but I plan to look into it soon.

**Comment by kaxil** [comment]
Thanks Dhruv

**Comment by Lee-W** [comment]
> Right, so is my understanding correct 

Yes, this understanding is correct!

> Can you tell me the difference between AIR302 and AIR312 in the above proposal? 

AIR302 (original 303) is something like `airflow.api.auth.backend.basic_auth → airflow.providers.fab.auth_manager.api.auth.backend.basic_auth``. `AIR312` is for something like `airflow.sensors.filesystem.FileSensor → airflow.providers.standard.sensors.filesystem.FileSensor`. The only difference is the provider needed in `AIR312` is pre-installed with Airflow. So i'm also thinking of just merging it to AIR301 🤔 Before this decision is made. I can still work on other changes

**Comment by Lee-W** [comment]
> Can you help me in what order should I be reviewing the PRs? 

The regrouped has not yet been created (except for 301 to 002). Others are auto fix PRs. But after addressing all the comments on current PRs, I'll list them down here again. Thanks!

> Do you want to regroup the rules first? 

I'll create PRs today or tomorrow. We can look into the autofix first.

> Does that require changing some of the symbols that are currently being detected as hard deprecation but need to be moved to soft deprecations?

Yes, it will.

**Comment by Lee-W** [comment]
@dhruvmanila 

Would be great if I can get some feedback on https://github.com/astral-sh/ruff/pull/16965 first

The following four items should be easier to review, as some of them have already been approved.

https://github.com/astral-sh/ruff/pull/16976
https://github.com/astral-sh/ruff/pull/16977
https://github.com/astral-sh/ruff/pull/16978
https://github.com/astral-sh/ruff/pull/17085

Thanks!

**Comment by dhruvmanila** [comment]
Thank you, just acknowledging this. I've added this to my todo list for today, I'll start looking into it in a couple of hours.

**Comment by Lee-W** [comment]
Hey @dhruvmanila , just a heads up. @sunank200 will help with the reorg and apply the autofix to other rules for the following few days. I'll be back next Monday. 

I'll send a sample reorg PR later today (branched out from https://github.com/astral-sh/ruff/pull/16978) so it would be easier for @sunank200 to pick up.

Thanks for helping out!

**Comment by Lee-W** [comment]
@dhruvmanila Thanks so so so much for helping out and merging so many PRs!

---

@sunank200 As I'm out till next Monday, here are things we can do as the next steps. 

## Fix and adding rules
1. Get https://github.com/astral-sh/ruff/pull/17123 merged (This should be ready now. But please address the comments if there's any. Thanks!)
2. Address the comments on https://github.com/astral-sh/ruff/pull/17115#discussion_r2025052581 by expanding `Replacement::ImportPathMoved` to `Replacement::ProviderName` and get https://github.com/astral-sh/ruff/pull/17115 merged. This would also ease how AIR302 is fixed.

## Autofix
* Apply the `Replacement::AutoImport` thing to `Replacement::Name` in `crates/ruff_linter/src/rules/airflow/rules/removal_in_3.rs` whenever it makes sense (take a look at https://github.com/astral-sh/ruff/pull/16965) (https://github.com/astral-sh/ruff/pull/16965#pullrequestreview-2736632433)
* Fix https://github.com/astral-sh/ruff/pull/17038 and get it merged (I just found out some of the fix is not working, but I don't have enough time to get it fixed)
*  add try catch guard to method and class attribute as https://github.com/astral-sh/ruff/pull/16965#discussion_r2025037654 suggested

## Reorg
* Reorganized AIR301, AIR302 into (AIR301, AIR302, AIR311, AIR312 as listed down on https://github.com/apache/airflow/issues/48560#issuecomment-2772608006 and https://github.com/apache/airflow/issues/48560#issuecomment-2772747938. I have a super early draft PR https://github.com/astral-sh/ruff/pull/17152. Not sure whether it helps, but you use it or create a new one.

**Comment by dhruvmanila** [comment]
Just a heads up, @ntBre is going to be taking over looking at the remaining parts of the Airflow work. Feel free to tag me if there's anything specific that requires my input / context.

**Comment by ntBre** [comment]
Thanks for the list! I approved 1-3 with very minor comments, and I'll try to tackle 4 this evening or tomorrow morning.

**Comment by Lee-W** [comment]
Hey @ntBre , I'll try to use this issue to communicate if there's more than one PR that needs your help, so that you won't need to switch between PRs. 🙂 

---

* https://github.com/astral-sh/ruff/pull/17310 (this is approved, addressed all the comments, and rebased from the latest main)
    * https://github.com/astral-sh/ruff/pull/17422 is based on the PR above. It adds a few crucial checks to AIR311 without other major changes. @sunank200 will help rebase once #17310 is merged


It would be really nice if we could have them in 0.11.6 🙂 

---

I'll other autofixes things once I'm back next week.

**Comment by ntBre** [comment]
I'll review these today!

**Comment by Lee-W** [comment]
Hey @ntBre , 

Just a quick update on the PRs I have today

1. https://github.com/astral-sh/ruff/pull/17355
2. https://github.com/astral-sh/ruff/pull/17542
    1. https://github.com/astral-sh/ruff/pull/17553
3. https://github.com/astral-sh/ruff/pull/17571
4. https://github.com/astral-sh/ruff/pull/17570

**Comment by Lee-W** [comment]
Hey @ntBre ,

Another quick update on the PRs we need your help 🙂 
(The indentation indicates PR dependencies)

1. https://github.com/astral-sh/ruff/pull/17542
    1. https://github.com/astral-sh/ruff/pull/17553
        1. https://github.com/astral-sh/ruff/pull/17570
2. https://github.com/astral-sh/ruff/pull/17571
    1. https://github.com/astral-sh/ruff/pull/17598

**Comment by ntBre** [comment]
Thanks for these lists! I reviewed the top-level PRs and can move on to the next layer soon.

**Comment by Lee-W** [comment]
Hey @ntBre ,

Another quick up on the PRs we have now 🙂
(The indentation indicates PR dependencies)


1. https://github.com/astral-sh/ruff/pull/17553
    1. https://github.com/astral-sh/ruff/pull/17570
2. https://github.com/astral-sh/ruff/pull/17571
    1. https://github.com/astral-sh/ruff/pull/17598

**Comment by ntBre** [comment]
Thanks, I'll try to review these in the next couple of days (hopefully tomorrow).

I'm not personally familiar with that `remove_unused_imports` API, but it certainly sounds promising! (Just to recap, from reading the linked thread, the issue is that you're getting a `ConflictingName` error from `get_or_import_symbol` so you need to remove the existing import first?) I'm happy to try playing with it if you run into issues, but following that type-checking rule's example looks like it makes sense to me.

**Comment by Lee-W** [comment]
> Just to recap, from reading the linked thread, the issue is that you're getting a ConflictingName error from get_or_import_symbol so you need to remove the existing import first?)

Yes, that's what I'm thinking of. Otherwise, we'll have one of the imports shadowed and removed by `unused import` rules. Glad to hear it sounds reasonable. I failed yesterday, but will continue trying today 💪

**Comment by Lee-W** [comment]
Hey @ntBre ,

Just a quick update on what we have now 🙂 

1. https://github.com/astral-sh/ruff/pull/17913
2. https://github.com/astral-sh/ruff/pull/17887
3. https://github.com/astral-sh/ruff/pull/17884
4. https://github.com/astral-sh/ruff/pull/17946
5. https://github.com/astral-sh/ruff/pull/17941
    1. https://github.com/astral-sh/ruff/pull/17940 
7. https://github.com/astral-sh/ruff/pull/17942

I'm still working on enabling import removal. But to lay down the foundation, I created a few refactoring PR and minor issues I found during refactoring. 

Thanks!

**Comment by Lee-W** [comment]
Hey @ntBre ,

Just a quick update on what we have now 🙂 

1. https://github.com/astral-sh/ruff/pull/17913
2. https://github.com/astral-sh/ruff/pull/17887
3. https://github.com/astral-sh/ruff/pull/17884
4. https://github.com/astral-sh/ruff/pull/17941
    1. https://github.com/astral-sh/ruff/pull/17985
5. https://github.com/astral-sh/ruff/pull/17942
    1. https://github.com/astral-sh/ruff/pull/17940  

I'm still working on enabling import removal. But to lay down the foundation, I created a few refactoring PR and minor issues I found during refactoring. 

Thanks!

**Comment by ntBre** [comment]
Thanks. I reviewed (1), just waiting for the release process to finish today to merge. I also reviewed (2), approved with one question for you. I looked at (3) again, and Micha and I are still hoping we might be able to offer a fix there. I might play with this a bit myself today to see if I can understand the problem better.

I briefly looked at (4) and (5), but they sound closely related to (3). Is that true, or should I go ahead and review them before (3) is resolved?

**Comment by Lee-W** [comment]
> Thanks. I reviewed (1), just waiting for the release process to finish today to merge. I also reviewed (2), approved with one question for you. I looked at (3) again, and Micha and I are still hoping we might be able to offer a fix there. I might play with this a bit myself today to see if I can understand the problem better.
> 
> I briefly looked at (4) and (5), but they sound closely related to (3). Is that true, or should I go ahead and review them before (3) is resolved?

Hi, thanks for helping out! I'll take a look at and see how I can address `(2)` next Monday. I don't have a clear answer for how to fix best `(3)`. I still don't think it's the best idea to fix it wrongly so I'll mark it as a draft and return to it later.

`(4)` and `(5)` are not related to `(3)`, but more related to the auto fix thing I'm going to focus on next week. Also, the following PRs based on those 2 come with fixes to reflect the latest Airflow 3.0 implementation

**Comment by Lee-W** [comment]
Hey @ntBre ,

Just a quick update on what we have now 🙂 Those draft PRs are actually ready but depend on other PRs


1. https://github.com/astral-sh/ruff/pull/17887
2. https://github.com/astral-sh/ruff/pull/17941
    1. https://github.com/astral-sh/ruff/pull/17985
3. https://github.com/astral-sh/ruff/pull/17942
    1. https://github.com/astral-sh/ruff/pull/17940  

Thanks!

**Comment by Lee-W** [comment]
Hey @ntBre ,

Don't really have a lot of updates today. The following are the PRs ready to be reviewed 🙂 . I think I have some progress on the autofix thing. might be able to put up a draft tomorrow!


1. https://github.com/astral-sh/ruff/pull/17941
    1. https://github.com/astral-sh/ruff/pull/17985
2. https://github.com/astral-sh/ruff/pull/17942
    1. https://github.com/astral-sh/ruff/pull/17940  

Thanks!

**Comment by Lee-W** [comment]
@ntBre  Sure! I just rebased it!

**Comment by Lee-W** [comment]
Hey @ntBre ,

we have 2 pending PRs today

1. https://github.com/astral-sh/ruff/pull/18093 (would be nice if we can have this reviewed first. it currently applies only to AIR302 for easier review. but already extracted the functions to airflow.helper. should not be to difficult to port it to AIR301, 311 312)
2. https://github.com/astral-sh/ruff/pull/18215 (minor improvement)

**Comment by Lee-W** [comment]
Hey @ntBre , just a gentle ping. Thanks 🙂

**Comment by MichaReiser** [comment]
@ntBre is out for the week. He'll be back early next week

**Comment by Lee-W** [comment]
> @ntBre is out for the week. He'll be back early next week

Got it. Thanks so much for updating 🙂

**Comment by Lee-W** [comment]
Hey @ntBre , as https://github.com/astral-sh/ruff/pull/18093 has been approved. I just addressed the comments and applied them to other rules. 

* https://github.com/astral-sh/ruff/pull/18367
* https://github.com/astral-sh/ruff/pull/18363
* https://github.com/astral-sh/ruff/pull/18368

They're kept as drafts as they depend on https://github.com/astral-sh/ruff/pull/18093, but I think they won't conflict with each other and are ready now.

**Comment by Lee-W** [comment]
Thanks so much! I think this is the last major thing we'll have in a short while, but I'll still be around to make some minor improvement

**Comment by zach-overflow** [comment]
Hi all, let me know if I should create a separate issue for this, but I found that currently (as of `ruff` version `0.14.1`), `AIR311` does not raise any errors for DAG code with deprecated imports from `airflow.models.param`.

To be clear, I don't believe this is a regression, just something that was never caught by the `AIR*` rules before.

**Comment by Lee-W** [comment]
> Hi all, let me know if I should create a separate issue for this, but I found that currently (as of `ruff` version `0.14.1`), `AIR311` does not raise any errors for DAG code with deprecated imports from `airflow.models.param`.
> 
> To be clear, I don't believe this is a regression, just a something that was never caught by the `AIR*` rules before.

https://github.com/astral-sh/ruff/pull/21043

PR created

**Comment by dhruvmanila** [comment]
@ntBre @Lee-W should we close this issue now that the rules have been stabilized?

**Comment by ntBre** [comment]
Sounds good to me if it works for @Lee-W!

**Comment by Lee-W** [comment]
Yep, sounds good :) Even though we still have a few minor things to add, they've been tracked in our end.

**Comment by dhruvmanila** [comment]
Sounds good, I'll close this issue but feel free to open additional issues if you want to track additional work here.
